accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,221 @@
+
+## Today's use cases for SecureString
+
+> This list is largely drawn from the feedback at https://github.com/dotnet/runtime/issues/30612 plus our analysis of internal consumers.
+
+### Storing credentials
+
+Some users may use SecureString as a convenient way to store credentials for a remote server, such as a SQL server. This allows them to pass around a "credentials object" that has somewhat more foolproof usage patterns than handing around a simple string.
+
+For these scenarios we should strongly encourage users to use passwordless mechanisms, such as using X.509 certificates or platform-specific credential management. These provide better, more modern security guarantees than typical username + password combinations.
+
+For scenarios which continue to use username + password, we should encourage users to pass around a `NetworkCredential` object instead. The NetworkCredential class abstracts away how the password is represented in memory: it can be a string, a legacy SecureString, or any future mechanism we create. Encouraging users to pass around this container rather than a raw SecureString instance allows their code to take advantage of future improvements we make to this space, even while they continue to use username + password credentials.
+
+### Preventing secrets from getting into crash dumps
+
+Users might rely on the fact that the SecureString backing memory is enciphered (only on Windows) to prevent secrets from getting into crash dumps. This is not ideal for a variety of reasons. The secret won't be protected before it's converted into a SecureString or after it's converted from a SecureString (such as during a p/invoke call). `SocketsHttpHandler` [converts SecureString instances to normal strings](https://github.com/dotnet/runtime/blob/0e23ef47ff8a2de6a5b5898a849ba00303aba8df/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.cs#L176-L185) as part of its operation. WPF's `PasswordBox` control [contains code](https://github.com/dotnet/wpf/blob/53812c590c5f140297ed96ee2a78d05dffa7be35/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/PasswordTextContainer.cs#L74-L97) to manipulate SecureString instances from string-based input. `SqlClient` [copies SecureString data into normal managed arrays](https://github.com/dotnet/SqlClient/blob/ead806e87a15be9f54c99db2ebe8918025c20fc9/src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/Interop/SNINativeMethodWrapper.cs#L989-L993) during manipulation. In all of these cases the secret remains eligible for inclusion in a memory dump.
+
+That secrets originate as strings deserves a bit more attention. The secret almost certainly entered the application via an HTTP request, or it was read from config, or it was entered into a WinForms textbox. These copies of the secrets are still floating around in memory (potentially even in a shared pool!), and SecureString does nothing to mitigate their being leaked.
```

Comment:
Presumably this (that you don't control all the buffers) is also a potential criticism of the proposal to have GC reliably zero memory (https://github.com/dotnet/runtime/issues/10480) ?

Reply:
Correct. Having the GC eagerly zero memory relies on arrays not being pooled, etc. Documents that discuss hypothetical "high security" modes for the runtime often hit on this point as well, saying that a good end-to-end solution also involves adding Kestrel flags which prohibit pooled array usage, etc.

Reply:
Such a mode should probably also zero memory going back into the array pool, as well as zeroing it before handing out the first time. Maybe we should have an issue for that - the closest is https://github.com/dotnet/runtime/issues/7532

Reply:
> That presumably would be much cheaper than not pooling at all.

Not as secure, though. The "avoid the pool" flag would be opt-in for applications which want better guarantees about avoiding potential use-after-free issues. Zeroing arrays when they're returned to the pool will not avoid this issue, unfortunately. :(